src/coreclr/jit/promotiondecomposition.cpp

```diff
@@ -24,13 +25,1181 @@ class DecompositionStatementList
+                    if (rep->NeedsWriteBack)
+                    {
+                        statements->AddStatement(
+                            Promotion::CreateWriteBack(m_compiler, m_src->AsLclVarCommon()->GetLclNum(), *rep));
+                        JITDUMP("  Will write back V%02u (%s) to avoid an additional write barrier\n", rep->LclNum,
+                                rep->Description);
+
+                        rep->NeedsWriteBack = false;
+                    }
+                }
+            }
+        }
+
+        GenTree*     addr       = nullptr;
+        GenTreeFlags indirFlags = GTF_EMPTY;
+
+        if (m_dst->OperIs(GT_BLK))
+        {
+            addr = m_dst->gtGetOp1();
+            indirFlags =
+                m_dst->gtFlags & (GTF_IND_VOLATILE | GTF_IND_NONFAULTING | GTF_IND_UNALIGNED | GTF_IND_INITCLASS);
+        }
+        else if (m_src->OperIs(GT_BLK))
+        {
+            addr = m_src->gtGetOp1();
+            indirFlags =
+                m_src->gtFlags & (GTF_IND_VOLATILE | GTF_IND_NONFAULTING | GTF_IND_UNALIGNED | GTF_IND_INITCLASS);
+        }
+
+        int numAddrUses = 0;
+
+        if (addr != nullptr)
+        {
+            for (int i = 0; i < m_entries.Height(); i++)
+            {
+                if (!IsHandledByRemainder(m_entries.BottomRef(i), remainderStrategy))
+                {
+                    numAddrUses++;
+                }
+            }
+
+            if (remainderStrategy.Type != RemainderStrategy::NoRemainder)
+            {
+                numAddrUses++;
+            }
+        }
+
+        bool needsNullCheck = false;
+        if ((addr != nullptr) && m_compiler->fgAddrCouldBeNull(addr))
+        {
+            switch (remainderStrategy.Type)
+            {
+                case RemainderStrategy::NoRemainder:
+                case RemainderStrategy::Primitive:
+                    needsNullCheck = true;
+                    // See if our first indirection will subsume the null check (usual case).
+                    for (int i = 0; i < m_entries.Height(); i++)
+                    {
+                        if (IsHandledByRemainder(m_entries.BottomRef(i), remainderStrategy))
+                        {
+                            continue;
+                        }
+
+                        const Entry& entry = m_entries.BottomRef(0);
+
+                        assert((entry.FromLclNum == BAD_VAR_NUM) || (entry.ToLclNum == BAD_VAR_NUM));
+                        needsNullCheck = m_compiler->fgIsBigOffset(entry.Offset);
+                        break;
+                    }
+                    break;
+            }
+        }
+
+        if (needsNullCheck)
+        {
+            numAddrUses++;
+        }
+
+        if ((addr != nullptr) && (numAddrUses > 1))
+        {
+            if (addr->OperIsLocal() && (!m_dst->OperIs(GT_LCL_VAR, GT_LCL_FLD) ||
+                                        (addr->AsLclVarCommon()->GetLclNum() != m_dst->AsLclVarCommon()->GetLclNum())))
```

Comment:
I think this is missing address exposed check, but it seems to be preexisting so let me fix it in a follow-up.